File: src/test/java/com/neo4j/data/importer/Neo4jDetectorIT.java
package com.neo4j.data.importer;

import static com.neo4j.data.importer.Neo4jVersion.V5_0_0;
import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;
import org.neo4j.driver.AuthTokens;
import org.neo4j.driver.Driver;
import org.neo4j.driver.GraphDatabase;
import org.testcontainers.containers.Neo4jContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;

@Testcontainers
class Neo4jDetectorIT {

    @Container
    private static final Neo4jContainer<?> enterprise = new Neo4jContainer<>("neo4j:5-enterprise")
            .withEnv("NEO4J_ACCEPT_LICENSE_AGREEMENT", "yes")
            .withAdminPassword("letmein!");

    @Container
    private static final Neo4jContainer<?> community = new Neo4jContainer<>("neo4j:5")
            .withEnv("NEO4J_ACCEPT_LICENSE_AGREEMENT", "yes")
            .withAdminPassword("letmein!");

    @Test
    void detects_neo4j_EE_instance() {
        try (Driver driver = GraphDatabase.driver(enterprise.getBoltUrl(), AuthTokens.basic("neo4j", "letmein!"))) {
            Neo4j neo4j = Neo4jDetector.detect(driver);

            assertThat(neo4j.edition()).isEqualTo(Neo4jEdition.ENTERPRISE);
            assertThat(neo4j.version()).isGreaterThanOrEqualTo(V5_0_0);
            assertThat(neo4j.environment()).isGreaterThanOrEqualTo(Neo4jEnvironment.ON_PREMISE);
        }
    }

    @Test
    void detects_neo4j_CE_instance() {
        try (Driver driver = GraphDatabase.driver(community.getBoltUrl(), AuthTokens.basic("neo4j", "letmein!"))) {
            Neo4j neo4j = Neo4jDetector.detect(driver);

            assertThat(neo4j.edition()).isEqualTo(Neo4jEdition.COMMUNITY);
            assertThat(neo4j.version()).isGreaterThanOrEqualTo(V5_0_0);
            assertThat(neo4j.environment()).isGreaterThanOrEqualTo(Neo4jEnvironment.ON_PREMISE);
        }
    }
}

File: neo4j-detection/src/test/kotlin/com/neo4j/data/importer/Neo4jDetectorIT.kt
package com.neo4j.data.importer

import com.neo4j.data.importer.DockerNeo4j.enterprise
import com.neo4j.data.importer.DockerNeo4j.image
import com.neo4j.data.importer.DockerNeo4j.version
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.neo4j.driver.AuthTokens
import org.neo4j.driver.GraphDatabase
import org.testcontainers.containers.Neo4jContainer
import org.testcontainers.junit.jupiter.Container
import org.testcontainers.junit.jupiter.Testcontainers

@Testcontainers
internal class Neo4jDetectorIT {
  @Test
  fun detects_neo4j_instance() {
    GraphDatabase.driver(neo4j.boltUrl, AuthTokens.basic("neo4j", "letmein!")).use { driver ->
      val neo4j = Neo4j.detectedWith(driver)
      assertThat(neo4j.edition)
          .isEqualTo(if (enterprise()) Neo4jEdition.ENTERPRISE else Neo4jEdition.COMMUNITY)
      // we cannot match more than the major since "5" is a valid tag
      assertThat(neo4j.version.major).isEqualTo(majorOf(version()))
      assertThat(neo4j.deploymentType).isGreaterThanOrEqualTo(Neo4jDeploymentType.SELF_MANAGED)
    }
  }

  private fun majorOf(version: String): Int {
    val dot = version.indexOf('.')
    if (dot == -1) {
      return version.toInt(10)
    }
    val major = version.substring(0, dot)
    return major.toInt(10)
  }

  companion object {
    @Container
    private val neo4j: Neo4jContainer<*> =
        Neo4jContainer(image())
            .withEnv("NEO4J_ACCEPT_LICENSE_AGREEMENT", "yes")
            .withAdminPassword("letmein!")
  }
}
